### DISCUSSION TITLE
关于 AsIs、IPIfNonMatch、IPOnDemand 的理解

### DISCUSSION THREAD

**Discussion by chshouyu** [XTLS/Xray-core#3310] [General]
> - `"AsIs"`：只使用域名进行路由选择。默认值。
> - `"IPIfNonMatch"`：当域名没有匹配任何规则时，将域名解析成 IP（A 记录或 AAAA 记录）再次进行匹配；
>   - 当一个域名有多个 A 记录时，会尝试匹配所有的 A 记录，直到其中一个与某个规则匹配为止；
>   - 解析后的 IP 仅在路由选择时起作用，转发的数据包中依然使用原始域名；
> - `"IPOnDemand"`：当匹配时碰到任何基于 IP 的规则，将域名立即解析为 IP 进行匹配；

上面是官方文档的定义，主要容易理解错的地方在于 IPIfNonMatch 和 IPOnDemand，根据我自己的试验，发现大概是这样子：

- IPIfNonMatch 会将域名规则和 IP 规则分成两个阶段，当域名规则阶段匹配不到时，将域名解析成 IP 再使用 IP 规则阶段
- IPOnDemand 则是遇到域名规则就使用域名规则，如果匹配则直接使用；遇到 IP 规则就将域名解析成 IP 然后使用 IP 规则匹配

换句话说，IPOnDemand 并不是不理会域名规则，而是它使用 IP 规则的时机要比 IPIfNonMatch 更及时一些，遇到什么规则就立即使用什么规则，这样理解对吗？
又或者说，IPOnDemand 才是真正意义上的从上到下依次匹配？

**Comment by dawncold** [comment]
我感觉你理解的正确。如果在路由规则中，把域名规则和IP规则完全分开，先域名后IP，那`IPIfNonMatch`和`IPOnDemand`将会是相同效果。可能得看源码验证。

**Comment by wtfr-dot** [comment]
我的理解是：文档里说的IPOnDemand策略是规则里只要有ip规则，则会首先解析出ip然后再进行匹配，所以每一次都要解析域名，这个解析需要向外查询，所以就慢了，如果设置有缓存，解析的域名在内部有ip映射，那么第二次再遇到同一个域名，就不需要再向外解析域名，直接从缓存里找，就有点相当于AsIs了，但是AsIs虽然第一次不需要解析域名，如果域名不匹配要走兜底规则，还是要走解析这条路，只是这个解析不在路由阶段，而在出站其它模块去了，所以AsIs最快。

**Comment by chise0713** [comment]
https://t.me/projectXray/3568429

**Comment by liweishu** [comment]
是的，无论从名字还是解释上来看 IPOnDemand 就是从上到下匹配，一旦遇到 ip 规则，立即进行将 domain 通过 dns 解析为 ip 进行匹配，再不匹配才往下组规则匹配。
IPIfNonMatch 则是 domain 规则和 ip 规则分开，domain 在 domain 规则不能匹配才会通过 dns 进行解析。
AsIs 下，domain 匹配不到就出站，不解析为 ip 进行匹配。
